moto/swf/models/__init__.py

```diff
@@ -0,0 +1,405 @@
+
+    def _get_domain(self, name, ignore_empty=False):
+        matching = [domain for domain in self.domains if domain.name == name]
+        if not matching and not ignore_empty:
+            raise SWFUnknownResourceFault("domain", name)
+        if matching:
+            return matching[0]
+        return None
+
+    def _check_none_or_string(self, parameter):
+        if parameter is not None:
+            self._check_string(parameter)
+
+    def _check_string(self, parameter):
+        if not isinstance(parameter, six.string_types):
+            raise SWFSerializationException(parameter)
+
+    def _check_none_or_list_of_strings(self, parameter):
+        if parameter is not None:
+            self._check_list_of_strings(parameter)
+
+    def _check_list_of_strings(self, parameter):
+        if not isinstance(parameter, list):
+            raise SWFSerializationException(parameter)
+        for i in parameter:
+            if not isinstance(i, six.string_types):
+                raise SWFSerializationException(parameter)
+
+    def _process_timeouts(self):
+        for domain in self.domains:
+            for wfe in domain.workflow_executions:
+                wfe._process_timeouts()
+
+    def list_domains(self, status, reverse_order=None):
+        self._check_string(status)
+        domains = [domain for domain in self.domains
+                   if domain.status == status]
+        domains = sorted(domains, key=lambda domain: domain.name)
+        if reverse_order:
+            domains = reversed(domains)
+        return domains
+
+    def register_domain(self, name, workflow_execution_retention_period_in_days,
+                        description=None):
+        self._check_string(name)
+        self._check_string(workflow_execution_retention_period_in_days)
+        self._check_none_or_string(description)
+        if self._get_domain(name, ignore_empty=True):
+            raise SWFDomainAlreadyExistsFault(name)
+        domain = Domain(name, workflow_execution_retention_period_in_days,
+                        description)
+        self.domains.append(domain)
+
+    def deprecate_domain(self, name):
+        self._check_string(name)
+        domain = self._get_domain(name)
+        if domain.status == "DEPRECATED":
+            raise SWFDomainDeprecatedFault(name)
+        domain.status = "DEPRECATED"
+
+    def describe_domain(self, name):
+        self._check_string(name)
+        return self._get_domain(name)
+
+    def list_types(self, kind, domain_name, status, reverse_order=None):
+        self._check_string(domain_name)
+        self._check_string(status)
+        domain = self._get_domain(domain_name)
+        _types = domain.find_types(kind, status)
+        _types = sorted(_types, key=lambda domain: domain.name)
+        if reverse_order:
+            _types = reversed(_types)
+        return _types
+
+    def register_type(self, kind, domain_name, name, version, **kwargs):
+        self._check_string(domain_name)
```

Comment:
This isn't a huge deal, but it feels like this validation should be getting done in SWFResponse instead of in the model. We like to try to keep the models clean so they can be reused in other places (for example Cloudformation).